rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,541 @@
+   $ tctl bots add example --public-key id_ed25519.pub
+   $ tbot start identity --token=challenge:id_ed25519
+   ```
+
+   (In this example, `tctl bots add` creates a `challenge` token automatically,
+   much like a `token`-type token is created automatically today.)
+
+   The public key can be copied as needed, similar to SSH `authorized_keys` and
+   GitHub's SSH authentication. This is arguably more secure since no secret is
+   ever copied.
+
+   On startup, Auth issues a challenge to the bot which is solved with its
+   private key, and it receives a standard renewable identity.
+
+2. **Bind-on-join**: The `tbot` client is given a joining secret.
+
+   Example UX:
+   ```
+   $ tctl bots add example --join-method=challenge
+   The bot token: challenge:04f0ceff1bd0589ba45c1832dfc8feaf
+   This token will expire in 59 minutes.
+   $ tbot start identity --token=challenge:04f0ceff1bd0589ba45c1832dfc8feaf
+   ```
+
+   On `tbot` startup, a keypair is transparently generated and exchanged with
+   Auth, after which the bot internally behaves as if flow 1 was used, and the
+   now-bound keypair perform its first full join.
+
+   From an end user's PoV, this process is nearly identical to traditional
+   `token` joining. While a joining secret does need to be copied to the bot
+   node, these secrets remain short-lived and one-time use.
+
+We expect most users to use Flow 2: it's much easier to provision new nodes and
+requires less back-and-forth between the admin's workstation and bot node. Flow
+1 is particularly ill-suited to Terraform use since keypairs would need to be
+pregenerated and copied to nodes, which is not ideal from a security PoV.
+
+#### Token Resource Example
+
+`challenge`-type tokens differ from other types in that they are intended to
+have no resource-level expiration (though that is allowed), are meant to have
+their spec modified over time by users or automation tools, and publish
+information about their current state in the immutable (to users) `status`
+field.
+
+``` yaml
+kind: token
+version: v2
+metadata:
+  name: my-join-token
+spec:
+  bot_name: example
+
+  join_method: challenge
+  challenge:
+
+    # `onboarding` parameters control initial join behavior
+    onboarding:
+      # If set, no joining secret is generated; the secret exchange ceremony is
+      # skipped and instance will directly prove its identity using its private
+      # key. It is an error for a public key to be associated with more than one
+      # token, and creation or update will fail if a public key is reused.
+      public_key: null
+
+      # If set, use an explicit initial joining secret; if both this and
+      # `public_key` are unset, a value will be generated server-side and
+      # stored in `.status.challenge.initial_join_secret`
+      initial_join_secret: ""
+
+      # Initial joining must take place before this timestamp. May be
+      # modified if bot has not yet joined.
+      expires: "2025-03-01T21:45:40.104524Z"
```

Comment:
Should we just use the token resource's expiry for this? Will it be confusing to have an expiry in both the metadata and the spec?

Reply:
It's separate because we don't want the resource to get purged after the expiration - the idea is that we can re-up a bot on the server by increasing its expiration/renewal count even after it's expired... which means the token resource itself needs to remain available.

Unfortunately that means if we want some time-based expiry layer that _doesn't_ lead to the resource getting garbage collected automatically, I think it needs to be in a separate field.

Reply:
I've renamed the two duplicate fields to `must_join_before` and `must_rejoin_before`, to hopefully reduce confusion.

Reply:
Bit of a nitpick: What about `can_rejoin_before` ? or `can_rejoin_until`? "Must" makes it sound like the Bot must rejoin.

Reply:
I've renamed the field to `may_join_until` (and standardized naming around "join" instead of "rejoin")